miner/collator.go

```diff
@@ -0,0 +1,243 @@
+// Copyright 2015 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package miner
+
+import (
+	"errors"
+	"math/big"
+	"sync/atomic"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core"
+	"github.com/ethereum/go-ethereum/core/state"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/params"
+)
+
+// BlockState represents a block-to-be-mined, which is being assembled. A collator
+// can add transactions, and finish by calling Commit.
+type BlockState interface {
+	// AddTransactions adds the sequence of transactions to the blockstate. Either all
+	// transactions are added, or none of them. In the latter case, the error
+	// describes the reason why the txs could not be included.
+	AddTransactions(sequence types.Transactions) error
+	// Commit signals that the collation is finished, and the block is ready
+	// to be sealed. After calling Commit, no more transactions may be added.
+	Commit()
+
+	Gas() (remaining uint64)
+	Coinbase() common.Address
+	BaseFee() *big.Int
+	Signer() types.Signer
+}
+
+// Collator is something that can assemble a block.
+type Collator interface {
+	CollateBlock(bs BlockState, pool Pool, interrupt *int32, isSealing bool) error
```

Comment:
Both `interrupt` and `isSealing` are for the recommit feedback. Can we somehow nuke them out from the plugin interface? So that the plugin implementers only need to take care of the transaction selection and ordering.

Actually after The Merge, the recommit is totally meaningless. Since in the PoS, we can only generate a single block for attestation.

Reply:
Regarding `isSealing`, I think it can be removed.  When not sealing, we can always use the standard strategy for constructing the pending block when the `newTxs` event fires.  So then `CollateBlock` only would get called during sealing.

Reply:
Regarding the interrupt.  I am thinking to abstract it inside `BlockState` so that the collator doesn't actually need to know/care about it (i.e. having `AddTransactions` fail when the interrupt goes off).

Reply:
Yup, looks good to me.